log/DESIGN.md

```diff
@@ -0,0 +1,674 @@
+Additionally, `Severity[Level]1` constants are defined to make the API more readable and user friendly.
+
+```go
+type Severity int
+
+const (
+	SeverityTrace1 Severity = iota + 1
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+
+	SeverityDebug1
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+
+	SeverityInfo1
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+
+	SeverityWarn1
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+
+	SeverityError1
+	SeverityError2
+	SeverityError3
+	SeverityError4
+
+	SeverityFatal1
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+
+	SeverityTrace = SeverityTrace1
+	SeverityDebug = SeverityDebug1
+	SeverityInfo  = SeverityInfo1
+	SeverityWarn  = SeverityWarn1
+	SeverityError = SeverityError1
+	SeverityFatal = SeverityFatal1
+)
+```
+
+[`SeverityText`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext)
+is accessed using following methods:
+
+```go
+func (r Record) SeverityText() string
+func (r *Record) SetSeverityText(s string)
+```
+
+[`Body`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+is accessed using following methods:
+
+```go
+func (r Record) Body() Value
+func (r *Record) SetBody(v Value)
+```
+
+[Log record attributes](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes)
+are accessed using following methods:
+
+```go
+func (r Record) WalkAttributes(f func(KeyValue) bool)
+func (r *Record) AddAttributes(attrs ...KeyValue)
+```
+
+`Record` has a `AttributesLen` method that returns
+the number of attributes to allow slice preallocation
+when converting records to a different representation:
+
+```go
+func (r Record) AttributesLen() int
+```
+
+Additionally, `Record` has a `Clone` method facilities
+coping a record with no shared state (attributes "back" slice):
+
+```go
+func (r Record) Clone() Record
+```
+
+The records attributes design and implemntation is based on
+[`slog.Record`](https://pkg.go.dev/log/slog#Record).
+It allows achieving high-performance access and manipulation of the attributes
+while keeping the API user friendly.
+It relieves the user from making his own improvements
+for reducing the number of allocations when passing attributes.
+
+The following defintions are implementing the abstractions
+described in [the specification](https://opentelemetry.io/docs/specs/otel/logs/#new-first-party-application-logs):
+
+```go
+type Value struct{}
+
+type Kind int
+
+const (
+	KindEmpty Kind = iota
+	KindBool
+	KindFloat64
+	KindInt64
+	KindString
+	KindBytes
+	KindList
+	KindMap
+)
+
+func (v Value) Kind() Kind
+
+func StringValue(value string)
+
+func IntValue(v int) Value
+
+func Int64Value(v int64) Value
+
+func Float64Value(v float64) Value
+
+func BoolValue(v bool) Value
+
+func BytesValue(v []byte) Value
+
+func ListValue(vs ...Value) Value
+
+func MapValue(kvs ...KeyValue) Value
+
+func (v Value) Any() any
+
+func (v Value) String() string
```

Comment:
Should we rename the methods to `As[Type]` so that `AsString` would panic if the kind is not `KindString` or do we prefer to have `String` and return a string representation of each Kind?

Reply:
Addressed 16a5fbf26643215e879e49b06eec8f7b446